grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy

```diff
@@ -0,0 +1,137 @@
+ * By using this software in any fashion, you are agreeing to be bound by
+ * the terms of this license.
+ * You must not remove this notice, or any other, from this software.
+ **/
+package org.pih.warehouse.data
+
+import org.pih.warehouse.importer.ImportDataCommand
+import org.pih.warehouse.product.ProductAssociation
+import org.pih.warehouse.product.Product
+import org.springframework.validation.BeanPropertyBindingResult
+
+
+class ProductAssociationDataService {
+
+    /**
+     * Validate product association
+     */
+    Boolean validateData(ImportDataCommand command) {
+        log.info "Validate data " + command.filename
+
+        command.data.eachWithIndex { params, index ->
+            ProductAssociation productAssociationInstance = new ProductAssociation(params)
+
+            // disable update product association on import
+            if (productAssociationInstance?.id) {
+                command.errors.reject("Row ${index + 1}: Cannot edit existing associations via import")
+                return
+            }
+
+            Product product = Product.get(productAssociationInstance.product?.id)
+            Product associatedProduct = Product.get(productAssociationInstance.associatedProduct?.id)
+
+            if (!product) {
+                command.errors.reject(
+                    "Row ${index + 1}: Product with code '${productAssociationInstance.product?.id}' does not exist"
+                )
+                return
+            }
+            if (!associatedProduct) {
+                command.errors.reject(
+                    "Row ${index + 1}: Product with code '${productAssociationInstance.associatedProduct?.id}' does not exist"
+                )
+                return
+            }
+            if (!product.active) {
+                command.errors.reject(
+                    "Row ${index + 1}: Product with code '${productAssociationInstance.product?.id}' is inactive"
+                )
+                return
+            }
+            if (!associatedProduct.active) {
+                command.errors.reject(
+                    "Row ${index + 1}: Product with code '${productAssociationInstance.associatedProduct?.id}' is inactive"
+                )
+                return
+            }
+            if (product == associatedProduct) {
+                command.errors.reject("Cannot associate a product with itself")
+                return
+            }
+
+            // Check for association duplicates
+            List<ProductAssociation> foundProductAssociations = ProductAssociation.findAllWhere([
+                    product             : productAssociationInstance.product,
+                    associatedProduct   : productAssociationInstance.associatedProduct,
+                    code                : productAssociationInstance.code,
+            ])
+            if (foundProductAssociations && foundProductAssociations.size() > 0) {
```

Comment:
Is there a case where we want to update a product association?

Reply:
No, as specified in the ticket product association import should not allow association updates, only creations